apps/ltn/src/logic/turn_restrictions.rs

```diff
@@ -0,0 +1,347 @@
+    all_r.extend(restricted_t);
+
+    for other_r in all_r {
+        all_pt.extend(map.get_r(other_r).get_thick_polygon().get_outer_ring().clone().into_points());
+    }
+
+    // TODO the `200` value seems to work for some cases. But it is arbitrary and there is no science
+    // behind its the value. Need to work out what is an appropriate value _and why_.
+    Polygon::concave_hull(all_pt, 200).unwrap_or(Polygon::dummy())
+}
+
+/// Returns all roads that are possible destinations from the given "from_road" where the turn is currently
+/// prohibited by a turn restriction.
+pub fn restricted_destination_roads(map: &Map, from_road_id: RoadID, i: Option<IntersectionID>) -> HashSet<RoadID> {
+    let candidate_roads = possible_destination_roads(map, from_road_id, i);
+
+    let from_road = map.get_r(from_road_id);
+    let mut restricted_destinations: HashSet<RoadID> = HashSet::new();
+        
+    for (restriction, target_r) in &from_road.turn_restrictions {
+        if *restriction == RestrictionType::BanTurns && candidate_roads.contains(target_r) {
+            restricted_destinations.insert(*target_r);
+        }
+    }
+    for (via, target_r) in &from_road.complicated_turn_restrictions {
+        if candidate_roads.contains(via) {
+            restricted_destinations.insert(*via);
+            restricted_destinations.insert(*target_r);
+        }
+    }
+    restricted_destinations
+}
+
+/// checks that an Intersection ID is connected to a RoadID. Returns `true` if connected, `false` otherwise.
+fn verify_intersection(map: &Map, r: RoadID, i: IntersectionID) -> bool {
+    let road = map.get_r(r);
+    return road.dst_i == i || road.src_i == i
+}
+
+/// Returns a HashSet of all roads which are connected by driving from RoadID.
+/// This accounts for oneway restrictions, but not turn restrictions. eg:
+/// 
+/// - If a oneway restriction on either the 'from_road' or the 'target_road' would prevent driving from
+/// source to destination, then 'target_road' it will NOT be included in the result.
+/// - If a turn restriction exists and is the only thing that would prevent driving from 'from_road' or the
+/// 'target_road', then the 'target_road' will still be included in the result.
+/// 
+/// `i` is Optional. If `i` is `Some` then, it must be connected to `from_r`. It is used to filter
+/// the results to return only the destination roads that connect to `i`.
+/// 
+// TODO highlighting possible destinations for complicated turns (at present both sections of existing
+// complicated_turn_restrictions are included). However possible future complicated turns are not detected.
+//
+// TODO Rework `possible_destination_roads()` and `restricted_destination_roads()` to a extra function that
+// returns a tupple `(permitted, opposing_oneway, restricted_turn)`
+pub fn possible_destination_roads(map: &Map, from_r: RoadID, i: Option<IntersectionID>) -> HashSet<RoadID> {
+
+    if i.is_some() && !verify_intersection(map, from_r, i.unwrap()){
```

Comment:
I personally think the more verbose would be readable:
```
if let Some(i) = i {
  if !verify_intersection(map, from, i) {
    panic
  }
}
```

Could also do `i.map(|i| !verify_intersection(map, from_i, i)).unwrap_or(false)`

Reply:
Agreed, but (possibly due to my relative inexperience in Rust) I dislike shadowing a variable when I'm going to use the original later in the function, so I've gone for this:

```
if let Some(unverified_i) = i {
    if !verify_intersection(map, from_r, unverified_i){
        panic!("IntersectionID {:?}, does not connect to RoadID {:?}", unverified_i, from_r);
    }
}
```

It's convenient to have `i` as an `Option<IntersectionID>` rather than `IntersectionID` later in the function.